libs/pika/async_mpi/include/pika/async_mpi/mpi_helpers.hpp

```diff
@@ -175,16 +133,49 @@ namespace pika::mpi::experimental::detail {
                 else
                 {
                     // pass the result onto a new task and invoke the continuation
-                    auto snd0 = ex::just(status) | ex::transfer(default_pool_scheduler(p)) |
-                        ex::then([receiver = PIKA_MOVE(receiver)](int status) mutable {
-                            PIKA_DETAIL_DP(
-                                mpi_tran<5>, debug(str<>("set_value_error_helper"), status));
-                            set_value_error_helper(status, PIKA_MOVE(receiver));
+                    auto snd0 = ex::transfer_just(default_pool_scheduler(p)) |
+                        ex::then([receiver = PIKA_MOVE(receiver)]() mutable {
+                            PIKA_DETAIL_DP(mpi_tran<5>, debug(str<>("set_value")));
+                            ex::set_value(PIKA_FORWARD(Receiver, receiver));
                         });
                     ex::start_detached(PIKA_MOVE(snd0));
                 }
             },
             request);
     }
 
+    // -----------------------------------------------------------------
+    // handler_method::continuation
+    // adds a request callback to the mpi polling code which will call
+    // the set_value/set_error helper using the void return signature
+    template <typename OperationState>
+    void add_continuation_request_callback(MPI_Request request, OperationState& op_state)
+    {
+        detail::add_request_callback(
+            [&op_state](int status) mutable {
+                PIKA_DETAIL_DP(mpi_tran<5>, debug(str<>("callback_void")));
+                set_value_error_helper(status, PIKA_MOVE(op_state.receiver));
+            },
+            request);
+    }
+
+    // -----------------------------------------------------------------
+    // handler_method::mpix_continuation
+    /// typedef int (MPIX_Continue_cb_function)(int rc, void *cb_data);
+    template <typename OperationState>
+    int mpix_callback([[maybe_unused]] int rc, void* cb_data)
+    {
+        PIKA_DETAIL_DP(mpi_tran<1>, debug(str<>("MPIX"), "callback triggered"));
+        auto& op_state = *static_cast<OperationState*>(cb_data);
+        // wake up the suspended thread
+        {
+            std::lock_guard lk(op_state.mutex);
+            op_state.status = rc;
+            op_state.completed = true;
+        }
+        op_state.cond_var.notify_one();
```

Comment:
I suspect that we need to protect this with the lock as well but details of lifetimes may mean that it's not necessary, please ignore if you think this is fine. The case I'm thinking about is this:

|  thread 1 | thread 2 |
| --- | --- |
| register continuation | |
| | continuation ready |
| | `mpix_callback` |
| | take lock |
| | set completed = true |
| | release lock |
| take lock | |
| wait with predicate | |
| completed is true, continue | |
| ??? | |
| release condition variable | |
| | notify condition variable |
| | :rotating_light: |

Is there anything stopping this from happening? This is obviously very unlikely but not impossible, and will happen sooner or later.

Reply:
currently this can't happen because I've used `MPIX_CONT_POLL_ONLY` which means that the continuation can only be called when we call `MPI_Test` and since that is protected by a lock in the multithreaded polling loop, and called on the thread that is setting the continuation in the single-threaded version, the continuatiob cannot be called until after we have set everything up correctly.
At some point I will upgrade the `mpix_continuation` version to allow the non `POLL_ONLY` callbacks, and also consider adding support for executing the sender continuation directly rather than using a suspend/resume approach. But to add this in, I will need a new PR

Reply:
Thanks, I see how this might not happen in the single-threaded version, but if there's no dedicated MPI pool and/or the MPI request is created inline (not on the MPI pool) then I don't see how the locks help. The lock in the polling loop is unrelated to the condition variable lock, no?

Reply:
condition_variable.wait checks the predicate 
```
            while (!pred()) { wait(lock); }
```
before actually waiting, but in  mpi_helper, the `completed = true` is done under the lock, so if thereread A sets the callback, thread B triggers the callback, then when thread A tries to wait - it will already have completed = true and it will skip the wait and we won't miss a `notify_one`
Am I mistaken?  (but you are correct, my initial explanation was incorreect they are not the same locks)

Reply:
The issue is not about missing a notify. Indeed the `wait` correctly checks the predicate, so it'll continue only after the continuation has been triggered. The issue is about calling `notify` on a condition variable that has been released.

Reply:
calling notify on a release cond_var is not a problem. nothing happens. what is bothering you about it? I can move the set callback function inside the lock if it makes you happy.

Reply:
Exactly! Sorry, I updated "release" to "destroy" in the table in the first comment.

Reply:
Note, https://github.com/pika-org/pika/pull/1128/commits/53f257eee822cf9a874059eafe78d5f8407df9b2 does not fix the issue, and I would remove that commit. The registering of a continuation does not need to happen under a lock. The notification of the condition variable has to be done under a lock (the code I'm commenting on here).

Reply:
It does fix it surely, because notify cannot be called until the lock is released, and that only happens when the `cond_var` goes into wait. At this point the `cond_var` is guaranteed  to be alive.  
Previously, registering the continuation was outside the lock, it could be triggered and `cond_var` notified by thread 2 in your table above `before` the lock was taken by thread 1, and the `cond_var` might be destroyed before thread 2 calls notify.  

New table


|  thread 1 | thread 2 |
| --- | --- |
| take lock | |
| register continuation | |
| | continuation ready (earliest possible) |
| | `mpix_callback` |
| wait with predicate (+unlock)| |
| | take lock |
| | set completed = true |
| | release lock |
| | notify *destroyed* condition variable |
| wake - completed is true, continue | |
| destroy condition variable | |
| | :smile: : no segfaults possible |

I'm not keen on calling `notify` under lock, because that guarantees contention on the wakeup. Is my analysis faulty?

Reply:
I think the unlock of this internal will still be problematic: https://github.com/pika-org/pika/blob/a40b9b5fac72969c5fd99942a3ce169b92d299aa/libs/pika/synchronization/include/pika/synchronization/condition_variable.hpp#L68. The notify will wake up the waiting thread, the waiting thread will proceed to destroy the condition variable, and unlocking the internal mutex will be a write to freed memory.

Not directly related to this PR, but still relevant: I've started looking into running the unit tests with _all_ the completion modes. Once we have that we would also get runs with thread sanitizer on the MPI continuations mode, and we can then a bit more confidently try to relax constraints if it looks like it's needed and/or possible.

There may of course also be ways of rewriting the condition variable itself to avoid issues like that. However, until then I'd prefer to go with the safer option.

How does that sound?

Reply:
If the condition variable can only be woken up when the notifier holds the lock, then something is very wrong. 

The line you highlighted is where a lock is taken, which is moved into the `detail::cond_var` notify_one and from there, a task is taken from the queue and resume is called on it - if that task were now to destroy the `cond_var` immediately - the one thing that could go wrong is that the next operation - unlocking the lock as it goes out of scope - an  `atomic<bool>` stored in the cond_var_data mutex -- is set to false. If the data is already destroyed, that atomic set will fail. 
Is it possible to have the condition variable destroy itself whilst I'm still holding that lock with its internal mutex?
If tjhjere is nothing that checks that lock before the destructor is called, then there might be a problem. I looked through the code and can't see anything protecting it. 
Since the `cond_var` data is held with an `intrusive_ptr` then one could m ake it safe by adding an `inclusive_ptr_add_ref` inside an RAII wrapper inside the top level notify - then the data could not go out of scope if the woken thread deletes the cond_var.

This seems wrong. what have I misunderstood?

Reply:
> If the condition variable can only be woken up when the notifier holds the lock, then something is very wrong.

I kind of agree, but I'm not confident enough about the implementation to say if it has to be this way or if there's a way around it. There are cases where it'd be safe to do the notify without a lock, such as if both waiter and notifier hold a reference count on a heap-allocated condition variable.

> The line you highlighted is where a lock is taken, which is moved into the `detail::cond_var` notify_one and from there, a task is taken from the queue and resume is called on it - if that task were now to destroy the `cond_var` immediately - the one thing that could go wrong is that the next operation - unlocking the lock as it goes out of scope - an `atomic<bool>` stored in the cond_var_data mutex -- is set to false. If the data is already destroyed, that atomic set will fail. Is it possible to have the condition variable destroy itself whilst I'm still holding that lock with its internal mutex? If tjhjere is nothing that checks that lock before the destructor is called, then there might be a problem. I looked through the code and can't see anything protecting it.

Aha, I think I see now. Since the waiter has to re-take the internal lock after waking up, the notifier will release the lock before the condition variable is destroyed. With this I think this is indeed now safe...

Since the `cond_var` data is held with an `intrusive_ptr` then one could m ake it safe by adding an `inclusive_ptr_add_ref` inside an RAII wrapper inside the top level notify - then the data could not go out of scope if the woken thread deletes the cond_var.
> 
> This seems wrong. what have I misunderstood?

That could probably help, assuming the analysis above is wrong.

Sorry for being difficult about this. I think you're right that the latest version is ok as is. Unless I've now triggered you to figure out other ways that this can break, let's go with the current version.